CMakeLists.txt

```diff
@@ -431,42 +431,42 @@ endif()
 # but for back compatability, we need to treat enable-netcdf4
 # as equivalent to enable-hdf5.
 # We detect this using these rules:
-# 1. if ENABLE_HDF5 is off then disable hdf5
-# 2. if ENABLE_NETCDF4 is off then disable hdf5
+# 1. if NETCDF_ENABLE_HDF5 is off then disable hdf5
+# 2. if NETCDF_ENABLE_NETCDF4 is off then disable hdf5
 # 3. else enable hdf5
-option(ENABLE_NETCDF_4 "Use HDF5." ON)
-option(ENABLE_NETCDF4 "Use HDF5." ON)
-option(ENABLE_HDF5 "Use HDF5." ON)
-if(NOT ENABLE_HDF5 OR NOT ENABLE_NETCDF4 OR NOT ENABLE_NETCDF_4)
-  set(ENABLE_HDF5 OFF CACHE BOOL "Use HDF5" FORCE)
-endif()
-option(ENABLE_HDF4 "Build netCDF-4 with HDF4 read capability(HDF4, HDF5 and Zlib required)." OFF)
-option(ENABLE_DAP "Enable DAP2 and DAP4 Client." ON)
-option(ENABLE_NCZARR "Enable NCZarr Client." ON)
-option(ENABLE_PNETCDF "Build with parallel I/O for CDF-1, 2, and 5 files using PnetCDF." OFF)
-set(ENABLE_CDF5 AUTO CACHE STRING "AUTO")
-option(ENABLE_CDF5 "Enable CDF5 support" ON)
+option(NETCDF_ENABLE_NETCDF_4 "Use HDF5." ON)
+option(NETCDF_ENABLE_NETCDF4 "Use HDF5." ON)
+option(NETCDF_ENABLE_HDF5 "Use HDF5." ON)
```

Comment:
Removed. Added fatal error if using `NETCDF_ENABLE_NETCDF4`. `NETCDF_ENABLE_NETCDF_4` was used in more places and `NETCDF_ENABLE_NETCDF4` was only used in the top level cmake

Reply:
Where are the usage of `NETCDF_ENABLE_NETCDF_4`?

Reply:
Sorry, I meant `ENABLE_NETCDF4`. I'm not entirely sure what the use of it is

Reply:
@WardF? What is the intent with these options? The help messages seems to be problematic at the very least

Reply:
The situation is this.
Before the zarr code was added (libnczarr),  only the HDF5 code used libsrc4.
So up until zarr, the ENABLE_LIBSRC4 was set if and only if hdf5 was enabled.
After adding zarr, it was necessary to distinguish the use of libsrc4 from enabling hdf5.
So I added an explicit ENABLE_HDF5 option to decide if hdf5 was enabled,
and deprecated the use of ENABLE_LIBSRC4 for that purpose.
Now, if ENABLE_NCZARR and/or ENABLE_HDF5 was set, ENABLE_LIBSRC4 was implicitly enabled.

The one complication was that I knew it would take a while before everyone switched
to using ENABLE_HDF5, so I provided a way to allow ENABLE_LIBSRC4 to continue to imply
ENABLE_HDF5, and hence enable HDF5 use.

You can see this more clearly in configure.ac at about line 134  starting where it says
>\# An explicit disable of netcdf-4 | netcdf4 is treated as if it was disable-hdf5...

Reply:
@DennisHeimbigner, the issue is that it is hard to figure out the intent from the implementation, e.g.:
https://github.com/Unidata/netcdf-c/blob/7babdcb299313f4a46b19793d11e5a09c824dbb8/CMakeLists.txt#L489-L503

And we don't have the expertise to navigate the autotools implementation to mirror it, so we would need more back-and-forth and discussion about these options to properly converge to a more simpler cmakey solution. If possible, it would be nice if we can regex find all of the `option` and `cache` variables, put them in a table and figure out if each one is:
- standalone/top-level and  `cmake_dependent_option` options: These should be renamed and organized
- meta option that overwrites other options: These should be migrated to cmake preset, but can be designed with `set(FORCE)` + relevant messages as intermediary option (preferably with deprecation message in favor of the preset counterpart) 
- deprecated options: Removed in this cycle with appropriate deprecation messages
- complex dependent options (those that cannot be handled by `cmake_dependent_option`): To be reviewed together

For starter having just the list of it and column to mark in which category it belongs will help keep track and discuss these options.

For the deprecated options, on both cmake and autotools, could a message be added on the version of `hdf5` where such options are advertised to be removed. Judging by the release schedule so far, 1-2 minor version announcement period should be prudent enough.